### DISCUSSION TITLE
Clarifications regarding React Router's long-term vision

### DISCUSSION THREAD

**Discussion by ernestostifano** [remix-run/react-router#10496] [Q&A]
Hello everyone, just wanted to understand better your long-term vision before making any decision on big projects.

Under [Picking a Router](https://reactrouter.com/en/main/routers/picking-a-router#picking-a-router), it can be read:
> We recommend updating your app to use one of the new routers from 6.4.

But right after that, under [Web Projects](https://reactrouter.com/en/main/routers/picking-a-router#web-projects), there is:
> If you're not interested in the data APIs, you can continue to use `<BrowserRouter>`

So, the motives of your recommendation are not fully clear:
- Is it because new routers are going to eventually replace the old ones?
- Are you particularly focusing on the new routers, so new features are expected to be developed on top of the new routers?
- How do they compare in terms of performance and reliability/stability?
- **Or is it that you are suggesting people to use newer routers because they are based on the architecture principles of Remix?**

The later point is the one that worries me the most, especially if the true answer is a combination of the others too.

Probably because of how the documentation is written, or just because everyone likes to be up-to-date, people is trying really hard to migrate or to start new implementations using the new routers. The problem is that, despite some notorious benefits, the new routers seem to be a lot more opinionated than the old ones, almost forbidding any pattern that is slightly different from the proposed ones. They also include a lot of stuff that some people might consider "extras".

Just scratching the surface of the discussions I could find some perplexity that can be related to this:

- https://github.com/remix-run/react-router/discussions/10478
- https://github.com/remix-run/react-router/discussions/10223
- https://github.com/remix-run/react-router/discussions/10491
- https://github.com/remix-run/react-router/discussions/10166
- https://github.com/remix-run/react-router/discussions/9864
- https://github.com/remix-run/react-router/discussions/10440
- https://github.com/remix-run/react-router/discussions/10300
- https://github.com/remix-run/react-router/discussions/10309
- https://github.com/remix-run/react-router/discussions/10110

Don't get me wrong, Remix is really nice, but maybe not everyone wants to use it, it is one good pattern in a sea of other good ones, and sometimes the overhead is not acceptable.

**Maybe all of this is just a misunderstanding and it can be solved by modifying the documentation to clearly say: "Hey, you are free to use any of these routers based on your needs. Some of them focus on navigation only, while others add some Remix-style sugar, if that is your thing. All routers are stable, perform similarly and will be equally maintained in the foreseeable future."**

React Router is a major open source project, with almost 10 million downloads per week on NPM. It is safe to assume that the great majority of React apps out there are using some version of React Router. Please, be careful, because navigation is a core functionality and should be one of the less opinionated ones.

I am sure there are awesome features, patterns and ideas out there and it is great to make them widely available, but everything that is a plus, beyond classic navigation, should be opt-in, IMHO.

**Comment by AnielloFalcone** [comment]
Totally agree! Would be great to have some clarifications over the points above to better understand how to implement react-router in the best way.

**Comment by ryanflorence** [comment]
Thanks for the question! It's a good one.

The tl;dr is in your message already 😅

> I am sure there are awesome features, patterns and ideas out there and it is great to make them widely available, but everything that is a plus, beyond classic navigation, should be opt-in, IMHO.

Yep! That's why the docs say: 

> If you're not interested in the data APIs, you can continue to use `<BrowserRouter>`

We'll keep supporting that for the foreseeable future.

---

And for a bit more explanation, please continue 😁

> the new routers seem to be a lot more opinionated than the old ones, almost forbidding any pattern that is slightly different from the proposed ones

I don't think opinionated is the right word, but there are tradeoffs: things that make data loading more efficient work against things like "background location for modals". But you are right, there are some use cases that are solved well with the `<Routes>`-style matching that are currently unclear or cumbersome with `createBrowserRouter`. Some of that is because of us not spending the time to figure it out yet, and some of that is because of how React itself is changing. Like you, we're trying to figure out the direction our dependencies are going before we commit to APIs to meet those use cases! Fortunately, you can keep using `<Routes>` for them today and into the future.

I'd like to clarify something from this comment a bit:

> Or is it that you are suggesting people to use newer routers because they are based on the architecture principles of Remix?

Remix is based on the principles of React Router, not the other way around. Remix is a just a server and bundler for React Router.

In fact, v4/5 is the only version of React Router that **didn't** have some sort of data features. The original `v0.1.0` release had `willTransitionTo` and `willTransitionFrom` and `v1` had `onEnter` and `onLeave` to initiate data loading. Even v4 had a package `react-router-config` with examples for parallel data loading. There's a bit too much backstory to explain why we stopped going that direction with v4/5 directly, but with v6 (and Remix) we got back into it. In other words, data concerns have always been a part of React Router.

We added the new data features to Remix first because we weren't sure where the server, bundler, and router integration points were, so we just did it all where the bundler and server were. In an alternative timeline, I would have preferred to have done it in React Router first (the way new features are developed today) but it's difficult to know the end from the beginning.

In summary: you don't have to use every feature of React Router and we'll continue supporting matching components to URLs in the middle of the component tree for the foreseeable future. You can even mix and match. Use a `createBrowserRouter` at the top of the app with some routes with data loading, and render a `<Routes>` deeper in the tree for highly dynamic routing that you do today.

I hope this helps. Thanks for asking 🤘

(if you have any follow up questions, please let us know)
